openmeter/entitlement/adapter/entitlement.go

```diff
@@ -499,21 +499,127 @@ func (a *entitlementDBAdapter) UpdateEntitlementUsagePeriod(ctx context.Context,
 	return err
 }
 
-func (a *entitlementDBAdapter) ListActiveEntitlementsWithExpiredUsagePeriod(ctx context.Context, namespaces []string, expiredBefore time.Time) ([]entitlement.Entitlement, error) {
+func (a *entitlementDBAdapter) UpsertEntitlementCurrentPeriods(ctx context.Context, updates []entitlement.UpsertEntitlementCurrentPeriodElement) error {
+	return entutils.TransactingRepoWithNoValue(
+		ctx,
+		a,
+		func(ctx context.Context, repo *entitlementDBAdapter) error {
+			// Let's make sure there aren't any duplicate updates
+			uniqueUpdates := lo.UniqBy(updates, func(u entitlement.UpsertEntitlementCurrentPeriodElement) string {
+				return fmt.Sprintf("%s:%s", u.ID, u.Namespace)
+			})
+
+			if len(uniqueUpdates) != len(updates) {
+				return fmt.Errorf("%d duplicate entitlement updates provided", len(updates)-len(uniqueUpdates))
+			}
+
+			// We will check that all provided entitlements exist as we don't want to create any, just update the current usage period
+			entitlements, err := repo.db.Entitlement.Query().
+				// We're ignoring namespace here but as IDs are globally unique this should be fine
+				Where(db_entitlement.IDIn(slicesx.Map(updates, func(u entitlement.UpsertEntitlementCurrentPeriodElement) string {
+					return u.ID
+				})...),
+				).
+				All(ctx)
+			if err != nil {
+				return err
+			}
+
+			if len(entitlements) != len(updates) {
+				return fmt.Errorf("%d entitlement updates provided do not exist", len(updates)-len(entitlements))
+			}
+
+			// Create a map of entitlements for faster lookup
+			entMap := make(map[string]*db.Entitlement)
+			for _, ent := range entitlements {
+				entMap[ent.ID] = ent
+			}
+
+			// Now we will proceed with the update
+			dbUpdates := make([]*db.EntitlementCreate, 0, len(updates))
+			for _, update := range updates {
+				ent, ok := entMap[update.ID]
+				if !ok {
+					return fmt.Errorf("inconsistency error: entitlement %s not found", update.ID)
+				}
+
+				create := repo.db.Entitlement.Create().
+					// These fields will be ignored in the custom update
+					SetID(update.ID).
+					SetNamespace(ent.Namespace).
+					SetEntitlementType(ent.EntitlementType).
+					SetFeatureID(ent.FeatureID).
+					SetFeatureKey(ent.FeatureKey).
+					SetSubjectKey(ent.SubjectKey).
+					SetCurrentUsagePeriodStart(update.CurrentUsagePeriod.From).
+					SetCurrentUsagePeriodEnd(update.CurrentUsagePeriod.To)
+
+				dbUpdates = append(dbUpdates, create)
+			}
+
+			// Let's do a batch insert with on conflict do update
+			err = repo.db.Entitlement.CreateBulk(dbUpdates...).
+				OnConflict(
+					sql.ConflictColumns(db_entitlement.FieldID),
+					sql.ResolveWith(func(u *sql.UpdateSet) {
+						u.SetExcluded(db_entitlement.FieldCurrentUsagePeriodStart).
+							SetExcluded(db_entitlement.FieldCurrentUsagePeriodEnd)
+					})).Exec(ctx)
```

Comment:
```suggestion
             for _, batch := lo.Chunk(dbUpdates, (65535 / len(db_entitlement.Columns)) - 1) {
			err = repo.db.Entitlement.CreateBulk(batch...).
				OnConflict(
					sql.ConflictColumns(db_entitlement.FieldID),
					sql.ResolveWith(func(u *sql.UpdateSet) {
						u.SetExcluded(db_entitlement.FieldCurrentUsagePeriodStart).
							SetExcluded(db_entitlement.FieldCurrentUsagePeriodEnd)
					})).Exec(ctx)
				if err != nil {
				  return err
				  }
			}
```

There's no separate error.

Reply:
(Of course this is not commitable, I just made it up ;-) )

Reply:
I see your keenness on adding this, so I will, but just to iterate my opinion a paper trail again:

1. looking out for this is arbitrary: there are lots of system limitations we don't look out for mostly because we dont need to
2. i think we're in agreement that this scenario is misuse: we shouldn't be calling with this many updates so why allow for it?
3. the exact batching is arbitrary, its a "_safe_ upper bound", but still depends on the query. having an exact bound check is not trivial to do efficiently due to how ent works (without templating the query after each update)... `len(Columns)` isn't even necessarily an upper bound for all queries, e.g. when edges are set... but getting the exact param count is problematic:
```go
func getNumberOfQueryParamsFromQuery(q *db.EntitlementCreate) int {
	// Each EntitlementCreate query creates parameters for these fields if they're set
	// This is based on the fields in EntitlementMutation
	count := 0

	// Access the mutation directly to check which fields are set
	mutation := q.Mutation()

	// Required fields - always present in a create query
	// ID, namespace, entitlement_type, feature_id, feature_key, subject_key
	count += 6

	// Time fields - most creations will have these
	// created_at, updated_at
	count += 2

	// Optional fields - need to check if they're set
	if _, exists := mutation.Metadata(); exists {
		count++
	}
	if _, exists := mutation.DeletedAt(); exists {
		count++
	}
	if _, exists := mutation.ActiveFrom(); exists {
		count++
	}
	if _, exists := mutation.ActiveTo(); exists {
		count++
	}
	if _, exists := mutation.MeasureUsageFrom(); exists {
		count++
	}
	if _, exists := mutation.IssueAfterReset(); exists {
		count++
	}
	if _, exists := mutation.IssueAfterResetPriority(); exists {
		count++
	}
	if _, exists := mutation.IsSoftLimit(); exists {
		count++
	}
	if _, exists := mutation.PreserveOverageAtReset(); exists {
		count++
	}
	if _, exists := mutation.Config(); exists {
		count++
	}
	if _, exists := mutation.UsagePeriodInterval(); exists {
		count++
	}
	if _, exists := mutation.UsagePeriodAnchor(); exists {
		count++
	}
	if _, exists := mutation.CurrentUsagePeriodStart(); exists {
		count++
	}
	if _, exists := mutation.CurrentUsagePeriodEnd(); exists {
		count++
	}
	if _, exists := mutation.SubscriptionManaged(); exists {
		count++
	}

	// Also count any edge relationships that might be set
	// Note: We're not checking if edges are actually set to keep this simpler,
	// as they're rarely used in bulk operations for entitlements

	return count
}
```
4. this again opens deadlocks as (assuming everything is a conflict) we're practically acquiring EXCLUSIVE locks in 10k batches for the length of our transaction. if another transaction separately tries to lock 2 rows from two different batches in the unlucky order, we're deadlocked

With that in mind, I'll implement the exact form of this validation you suggested